File: apps/dsa-web/src/api/portfolio.ts
/**
 * Portfolio API Client - Holdings and watchlist management
 */

import { apiClient } from './client';

export interface HoldingItem {
  code: string;
  name?: string;
  assetType: 'stock' | 'fund' | 'etf' | 'hk_stock' | 'us_stock';
  shares: number;
  costBasis: number;
  currentPrice?: number;
  marketValue?: number;
  unrealizedPnl?: number;
  unrealizedPnlPct?: number;
  createdAt: string;
  updatedAt: string;
}

export interface HoldingsListResponse {
  total: number;
  stocks: HoldingItem[];
  funds: HoldingItem[];
  etfs: HoldingItem[];
  totalCost: number;
  totalMarketValue: number;
  totalPnl: number;
  totalPnlPct: number;
}

export interface CreateHoldingRequest {
  stock_code: string;
  shares: number;
  cost_basis: number;
}

export interface UpdateHoldingRequest {
  shares: number;
  cost_basis: number;
}

export interface WatchlistItem {
  code: string;
  name?: string;
  assetType: 'stock' | 'fund' | 'etf' | 'hk_stock' | 'us_stock';
  currentPrice?: number;
  changePercent?: number;
  createdAt: string;
}

export interface WatchlistResponse {
  total: number;
  stocks: WatchlistItem[];
  funds: WatchlistItem[];
  etfs: WatchlistItem[];
}

export interface AddWatchlistRequest {
  stock_code: string;
}

export const portfolioApi = {
  /**
   * Get all holdings with current prices and P&L
   */
  async getHoldings(assetType?: string): Promise<HoldingsListResponse> {
    const params = assetType ? { asset_type: assetType } : {};
    const response = await apiClient.get('/portfolio/holdings', { params });
    return response.data;
  },

  /**
   * Create a new holding or update if exists
   */
  async createHolding(data: CreateHoldingRequest): Promise<HoldingItem> {
    const response = await apiClient.post('/portfolio/holdings', data);
    return response.data;
  },

  /**
   * Update an existing holding
   */
  async updateHolding(code: string, data: UpdateHoldingRequest): Promise<HoldingItem> {
    const response = await apiClient.put(`/portfolio/holdings/${code}`, data);
    return response.data;
  },

  /**
   * Delete a holding
   */
  async deleteHolding(code: string): Promise<void> {
    await apiClient.delete(`/portfolio/holdings/${code}`);
  },

  /**
   * Get all watchlist items with current prices
   */
  async getWatchlist(assetType?: string): Promise<WatchlistResponse> {
    const params = assetType ? { asset_type: assetType } : {};
    const response = await apiClient.get('/portfolio/watchlist', { params });
    return response.data;
  },

  /**
   * Add item to watchlist
   */
  async addToWatchlist(data: AddWatchlistRequest): Promise<WatchlistItem> {
    const response = await apiClient.post('/portfolio/watchlist', data);
    return response.data;
  },

  /**
   * Remove item from watchlist
   */
  async removeFromWatchlist(code: string): Promise<void> {
    await apiClient.delete(`/portfolio/watchlist/${code}`);
  },
};

File: api/v1/schemas/portfolio.py
# -*- coding: utf-8 -*-
"""
Portfolio API Schemas - Pydantic models for request/response validation
"""

from typing import List, Optional
from pydantic import BaseModel, Field, field_validator

from data_provider.base import canonical_stock_code


class CreateHoldingRequest(BaseModel):
    """Request to create a new holding"""
    stock_code: str = Field(..., min_length=1, max_length=10, description="Stock/fund code")
    shares: float = Field(..., gt=0, description="Number of shares/units")
    cost_basis: float = Field(..., gt=0, description="Average cost per share")

    @field_validator('stock_code')
    @classmethod
    def validate_code(cls, v: str) -> str:
        return canonical_stock_code(v)


class UpdateHoldingRequest(BaseModel):
    """Request to update an existing holding"""
    shares: float = Field(..., gt=0, description="Number of shares/units")
    cost_basis: float = Field(..., gt=0, description="Average cost per share")


class HoldingItem(BaseModel):
    """Individual holding item with P&L calculation"""
    code: str
    name: Optional[str] = None
    asset_type: str
    shares: float
    cost_basis: float
    current_price: Optional[float] = None
    market_value: Optional[float] = None
    unrealized_pnl: Optional[float] = None
    unrealized_pnl_pct: Optional[float] = None
    created_at: str
    updated_at: str


class HoldingsListResponse(BaseModel):
    """Response with all holdings grouped by asset type"""
    total: int
    stocks: List[HoldingItem]
    funds: List[HoldingItem]
    etfs: List[HoldingItem]
    total_cost: float
    total_market_value: float
    total_pnl: float
    total_pnl_pct: float


class AddWatchlistRequest(BaseModel):
    """Request to add item to watchlist"""
    stock_code: str = Field(..., min_length=1, max_length=10, description="Stock/fund code")

    @field_validator('stock_code')
    @classmethod
    def validate_code(cls, v: str) -> str:
        return canonical_stock_code(v)


class WatchlistItem(BaseModel):
    """Individual watchlist item"""
    code: str
    name: Optional[str] = None
    asset_type: str
    current_price: Optional[float] = None
    change_percent: Optional[float] = None
    created_at: str


class WatchlistResponse(BaseModel):
    """Response with all watchlist items grouped by asset type"""
    total: int
    stocks: List[WatchlistItem]
    funds: List[WatchlistItem]
    etfs: List[WatchlistItem]
